storage/types.go

```diff
@@ -197,6 +199,7 @@ func NewNodeIDForTreeCoords(depth int64, index int64, maxPathBits int) (NodeID,
 }
 
 // SetBit sets the ith bit to true if b is non-zero, and false otherwise.
+// Note that the bit index 0 is the right most valid bit in the path.
```

Comment:
Consider the case when `NodeID` is 9 bits. According to the docstring of `PrefixLenBits`, `Path[1]` will have the MSB used as the 9th bit. When we count bits from the other side, this MSB should be the 0-th bit. But `SetBit(0, 1)` will change the **L**SB of `Path[1]`, which is not in the `NodeID`.

Reply:
Are we just lucky that our keys are multiples of 8?

Reply:
OK going to need to look at it again.

The multiple of 8 assumption is baked in pretty much throughout this code.

Reply:
Yeah, I think we should either throw away the code that assumes non-8 keys, or fix and test it.

Reply:
New plan is to write more tests against the current implementation and then use them against this. It's not worth taking chances.

Reply:
I think part of the confusion arises because Bit() and SetBit() are working with fixed 64-bit (for log trees) arrays whereas the node ID structure is a variable bit length and not necessarily byte aligned (for internal node references within subtrees). I think adding actual examples might make this clearer.

And I also need to avoid adding to the confusion by using terms like "to the right" without qualification.

Reply:
Also apparently node IDs mean [different things](https://github.com/google/trillian/blob/master/docs/storage/storage.md#nodeids) for maps and logs, so we should be twice as careful when reasoning about the bit index.

Reply:
Maps are "easier" as they're a 256 bit path in the notional sparse Merkle tree so don't do the same sort of bit twiddling we do for logs. But "harder" in that they have non uniform strata.